src/value.mts

```diff
@@ -22,149 +21,268 @@ import { EnvironmentRecord } from './environment.mjs';
+    Object.defineProperty(Value, 'undefined', { value: new this(), configurable: true });
+  }
+}
 
 /** https://tc39.es/ecma262/#sec-ecmascript-language-types-null-type */
-export class NullValue extends PrimitiveValue { }
+export class NullValue extends PrimitiveValue {
+  declare readonly type: 'Null';
+  declare readonly value: null;
+
+  private constructor() { // eslint-disable-line no-useless-constructor
+    super();
+  }
+
+  static {
+    Object.defineProperty(this.prototype, 'type', { value: 'Null', configurable: true });
+    Object.defineProperty(this.prototype, 'value', { value: null, configurable: true });
+    Object.defineProperty(Value, 'null', { value: new this(), configurable: true });
+  }
+}
 
 /** https://tc39.es/ecma262/#sec-ecmascript-language-types-boolean-type */
-export class BooleanValue extends PrimitiveValue {
-  readonly boolean: boolean;
-  constructor(v: boolean) {
+export class BooleanValue<T extends boolean = boolean> extends PrimitiveValue {
+  declare readonly type: 'Boolean';
+  readonly value: T;
+
+  private constructor(value: T) {
     super();
-    this.boolean = v;
+    this.value = value;
   }
 
   booleanValue() {
-    return this.boolean;
+    return this.value;
   }
 
   [Symbol.for('nodejs.util.inspect.custom')]() {
-    return `Boolean { ${this.boolean} }`;
+    return `Boolean { ${this.value} }`;
   }
-}
 
-Object.defineProperties(Value, {
-  undefined: { value: new UndefinedValue(), configurable: false, writable: false },
-  null: { value: new NullValue(), configurable: false, writable: false },
-  true: { value: new BooleanValue(true), configurable: false, writable: false },
-  false: { value: new BooleanValue(false), configurable: false, writable: false },
-});
-
-/** https://tc39.es/ecma262/#sec-ecmascript-language-types-string-type */
-class StringValue extends PrimitiveValue {
-  readonly string: string;
-  constructor(string: string) {
-    super();
-    this.string = string;
+  static {
+    Object.defineProperty(this.prototype, 'type', { value: 'Boolean', configurable: true });
+    Object.defineProperty(Value, 'true', { value: new this(true), configurable: true });
+    Object.defineProperty(Value, 'false', { value: new this(false), configurable: true });
   }
+}
 
-  stringValue() {
-    return this.string;
-  }
+/** https://tc39.es/ecma262/#sec-ecmascript-language-types-string-type */
+export interface StringValue extends InstanceType<typeof PrimitiveValue> {
+  readonly type: 'String';
+  readonly value: string;
+  stringValue(): string;
 }
-// rename for static semantics StringValue() conflict
-export { StringValue as JSStringValue };
+
+/**
+ * https://tc39.es/ecma262/#sec-static-semantics-stringvalue
+ */
+export const StringValue = (() => {
+  // NOTE: Using IIFE so that the class does not conflict with the type of the same name
+  /** https://tc39.es/ecma262/#sec-static-semantics-stringvalue */
+  @callable((_target, _thisArg, [value]) => {
```

Comment:
is there a reason you combined these? its a very strange combination considering it only exists because engine262's value classes happen to have "Value" in the name. This is why I originally renamed StringValue to JSStringValue.

Reply:
Mostly to avoid the conflict between the two. I ran into it a few times while looking at other type safety-related changes because it's still feasible to accidentally pull the wrong one in due to auto-imports in VS Code, and you already have similar overload behavior for `Value` and `Completion`, so it seemed like it made sense.

Reply:
What I'm trying to say is that these happen to have the same name by pure chance, there's no logical reason to combine them into one entity, that just seems confusing.

Reply:
True, though it's unfortunate that the naming scheme for `Value` causes a collision. I've reverted the change.